### PR TITLE
Support for non-u64 types for Window Bound

### LINKED ISSUE #3571: Hardcoded u64 for WindowFrameBound fields
# Hardcoded u64 for WindowFrameBound fields

**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

`WindowFrameBound` enum takes `Option<u64>` for `Preceding` and `Following` fields. When we run the below query in datafusion (valid in POSTGRESQL).

```sql
SELECT 
	COUNT(*) OVER(ORDER BY ts RANGE BETWEEN '1 DAY' PRECEDING AND '1 DAY' FOLLOWING) as cnt 
FROM t
```

it produces error below.

```rust
Error: SQL(ParserError("Expected literal int, found: '1 DAY'"))
```

**Describe the solution you'd like**

Instead of `Option<u64>`, we can use `Option<ScalarValue>` to support different types such as timestamps.

**Describe alternatives you've considered**

N.A

**Additional context**

The relevant code section can be found at the following link.

[arrow-datafusion/window_frame.rs at c3bf84e334d6a12c89bacec38e2baa02ca2417ac · apache/arrow-datafusion](https://github.com/apache/arrow-datafusion/blob/c3bf84e334d6a12c89bacec38e2baa02ca2417ac/datafusion/expr/src/window_frame.rs#L131)

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Support for non-u64 types for Window Bound

# Which issue does this PR close?

Closes #3571 . 

# Rationale for this change

With this change we are adding support for queries like below.

```sql
SELECT 
	COUNT(*) OVER(ORDER BY ts RANGE BETWEEN '1 DAY' PRECEDING AND '1 DAY' FOLLOWING) as cnt 
FROM t
```

# What changes are included in this PR?

- Closing this issue requires support for parsing expressions inside window frame queries. Hence we have updated sqlparser version.
- `assert_contains` and `assert_not_contains` macros are used in tests at different places. These macros were duplicated where they were used, we have moved them under [https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/test_util.rs](https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/test_util.rs) to prevent code duplication.
- Util codes for datetime arithmetic are moved under [https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/scalar.rs](https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/common/src/scalar.rs). They are reached now with `.add` and `.sub` api of `ScalarValue` enum.
- We removed type coercions between `ScalarValue` types. Arithmetic operations are done only between same types (for `datetime` types arithmetic operations are done only with `INTERVAL` types.) 
- Previously we were rejecting queries in the form

```sql
SELECT
      COUNT(c1) OVER (ORDER BY c2 RANGE BETWEEN 1 PRECEDING AND 2 PRECEDING)
      FROM aggregate_test_100
```

during window frame creation. Since we are supporting more reach variants such as 

```sql
SELECT 
	COUNT(*) OVER(ORDER BY ts RANGE BETWEEN '1 DAY' PRECEDING AND '1 MONTH' PRECEDING) as cnt 
FROM t
```

it is not easy to do rejection during creation. We now do calculations for whether window frame is valid after physical schema information is available. Where we can do validation calculations with appropriate types. This doesn’t change anything for end users.

- sqlparser had also changes in `Expr` type. We have incorporated those changes. However, possibly they are written with some new capabilities in mind. We just did the bare minimum to support the current test suit.

# Are there any user-facing changes?

N.A

### CODE REVIEW

File: datafusion/physical-expr/src/window/aggregate.rs
Comment (alamb) [inline datafusion/physical-expr/src/window/aggregate.rs:166]: 👍

File: datafusion/core/src/physical_plan/planner.rs
Comment (alamb) [inline datafusion/core/src/physical_plan/planner.rs:1375]: This operation is typically called "coercion" and is handled in datafusion for other expression types here: https://github.com/apache/arrow-datafusion/blob/master/datafusion/expr/src/type_coercion.rs#L18-L32 (and submodule).

Did you consider doing this conversion as part of coercion?

One reason to do it as part of the normal coercion is that then the proper types will be present for operations such as constant folding / constant propagation. This might allow for expressions like

```sql
                COUNT(*) OVER (ORDER BY ts RANGE BETWEEN INTERVAL '1' DAY + INTERVAL '1' DAY PRECEDING AND INTERVAL '3 DAY' FOLLOWING),
```

Eventually

**Review by alamb — COMMENTED** [review]
Thank you for the contribution @mustafasrepo  -- these PRs are very impressive; I am  very excited to see additional date/time functionality added to DataFusion and thus I try and prioritize reviewing such PRs.

One thing that would make it easier to review PRs such as this would be to break them up into multiple smaller PRs.  The benefits of doing so are:

1. decreases the contiguous focus time (hard to find some days) needed to review them
2. increases the number of people who can review / merge them (we have many reviewers / committers, but only a few who feel comfortable reviewing a PR that touches so much of the codebase)

For example in this case multiple PRs could be:

1. Update sqlparser-rs
2. Update parser and logical planning (but error if anything other than a uint64 was passed)
3. move the location of test utils /  code to common
5. physical planner / physical expr changes

The only question I have about this PR that I think needs to be addressed before merging is the type coercion (see comment inline)

cc @waitingkuo @avantgardnerio @ovr

File: datafusion/common/src/scalar.rs
Comment (avantgardnerio) [reply datafusion/common/src/scalar.rs:595]: The comment should be updated. I think we're waiting on `chrono:0.4.23` now...

**Comment by ozankabak** [comment]
Thank you Andrew for your careful review and comments (as always). We will go through them and update the PR soon.

File: datafusion/common/src/scalar.rs
Comment (mustafasrepo) [reply datafusion/common/src/scalar.rs]: @avantgardnerio I have updated the comment per your suggestion. Thanks

File: datafusion/expr/src/window_frame.rs
Comment (mustafasrepo) [reply datafusion/expr/src/window_frame.rs]: I have changed the error type to ParserError, thanks for pointing it out.

File: datafusion/expr/src/window_frame.rs
Comment (mustafasrepo) [reply datafusion/expr/src/window_frame.rs]: Changed per your suggestion. Thanks.

**Comment by alamb** [comment]
What is your thinking regarding moving coercion of out physical planning and into the coercion pass? Is that something you plan to do?

**Comment by ozankabak** [comment]
@mustafasrepo is trying to solidify his understanding of your suggestion. He will probably circle back tomorrow to make sure he gets what you meant right -- and then follow through if you guys are on the same page. He may ask for some clarifying directions from you on how to do this.

**Comment by mustafasrepo** [comment]
> What is your thinking regarding moving coercion of out physical planning and into the coercion pass? Is that something you plan to do?

Hi @alamb, I sent a commit moving coercion from physical planning to the optimizer [type_coercion.](https://github.com/synnada-ai/arrow-datafusion/blob/feature/window_bound_scalar/datafusion/optimizer/src/type_coercion.rs). Was this what you had in mind or I misunderstood what you meant? I can change the implementation according to your suggestions. Thanks for your feedback.

**Comment by alamb** [comment]
> Hi @alamb, I sent a commit moving coercion from physical planning to the optimizer [type_coercion](https://github.com/synnada-ai/arrow-datafusion/blob/b78eb162910e31cdee6b2747f1031c0514e65b96/datafusion/optimizer/src/type_coercion.rs#L416). Is this what you had in mind or I misunderstood what you meant? I can change the implementation according to your suggestions. Thanks for your feedback.

This is exactly right -- thanks

**Review by alamb — APPROVED** [review]
I think this PR is ready to go -- thank you @mustafasrepo .

I will plan to merge it tomorrow unless anyone else would like to offer suggestions or needs more time to review it

**Pull request merged**
